ServiceFabricProcessor preview

JamesBirdsall opened the pull request:
## Description

This PR brings the new ServiceFabricProcessor library, integrating Service Fabric and Event Hubs, along with a few necessary changes in the underlying client to support SFP.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [ ] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [ ] If applicable, the public code is properly documented.
- [ ] Pull request includes test coverage for the included changes.
- [ ] The code builds without any errors.

JamesBirdsall committed:
Initial checkin

Initial checkin of ServiceFabricProcessor

JamesBirdsall committed:
Change to receive handler, add metrics thread

Change from receive loop to receive handler so fabric shutdown will not be blocked by long receive timeouts.
Add thread to read metrics from user processor, still needs to be hooked up to fabric.

JamesBirdsall committed:
Hook up metrics reporting

Set up user metric in SF.
Create metric reporting thread to read metric from user code and pass to SF.

serkantkaraca committed:
Merge branch 'dev' into SFprocessor

JamesBirdsall committed:
Use fixed epoch number

Don't need to keep track of and increment an epoch number, just use 0. If there is more than one primary running for a partition it is because of a shutdown issue and it is desired for the new receiver to win.

JamesBirdsall committed:
Make CancellationToken more visible

Make CancellationToken an argument to certain user-implemented methods in IEventProcessor and IEventProcessorFactory. Not needed for closing (already cancelled) or error logging.

JamesBirdsall committed:
Partition ordinal is not partition id

Distinguish between the two properly.

JamesBirdsall committed:
Change when PartitionContext is created

Required due to the change in when partitionId becomes available. The partitionContext cannot be created until later, which means the IEventProcessor cannot be instantiated until later.

JamesBirdsall committed:
Merge branch 'SFprocessor' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Convert to normal directory structure, integrate with main project.

JamesBirdsall committed:
Add placeholder docs to make compiler happy

JamesBirdsall committed:
Remove metric setup code, change to multiple user metrics.

JamesBirdsall committed:
Fixes:
rename InitialOffsetProvider to InitialPositionProvider
start checkpoint manager after partition id acquired

JamesBirdsall committed:
Mock fixes, add argument to create receiver to help mocks

JamesBirdsall committed:
Missing bracket

JamesBirdsall committed:
Change name of event source

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
First pass at converting to composable model

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Remove unused delete method from ICheckpointMananger

JamesBirdsall committed:
Give SFP access to internals of base client

JamesBirdsall committed:
Implement runtimeinfo support

JamesBirdsall committed:
Some mocks not needed with access to client internals

JamesBirdsall committed:
Add shutdown notification option

JamesBirdsall committed:
Fix nullref in mock receiver

JamesBirdsall committed:
Project update

JamesBirdsall committed:
Dispose EventDatas after use
Temporarily add TestMode for size experiments

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Update to current Service Fabric

JamesBirdsall committed:
Changes required so SFP does not need internal access to client (#300)

* Changes required so SFP does not need internal access to client

* Avoid Mock function on EventData, allow direct setting

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Update versions for preview release

serkantkaraca commented on src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/Checkpoint.cs:
```diff
@@ -0,0 +1,151 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.using System;
+
+using System;
+using System.Collections.Generic;
```
Can you move all using into namespace section?

JamesBirdsall replied:
Done.

serkantkaraca commented on src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/ReliableDictionaryCheckpointMananger.cs:
```diff
@@ -0,0 +1,159 @@
+                        ConditionalValue<Dictionary<string, object>> rawCheckpoint = await
+                            this.store.TryGetValueAsync(tx, partitionId, Constants.ReliableDictionaryTimeout, cancellationToken);
+
+                        await tx.CommitAsync();
```
Do we need txn for get?

JamesBirdsall replied:
Yes, the transaction is a required argment to the get method.

serkantkaraca commented on src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/EventProcessorService.cs:
```diff
@@ -0,0 +1,488 @@
+
+        async Task IPartitionReceiveHandler.ProcessEventsAsync(IEnumerable<EventData> events)
+        {
+            if ((events != null) || ((events == null) && Options.InvokeProcessorAfterReceiveTimeout))
```
>if ((events != null) || ((events == null) && Options.InvokeProcessorAfterReceiveTimeout)) [](start = 12, length = 89)

Do you need this since receive handler will handle this condition? Just create an empty event set if events==null.

JamesBirdsall replied:
You are correct, no need to double-check since InvokeProcessorAfterReceiveTimeout is already hooked up to client receive pump.

serkantkaraca commented on src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/EventProcessorService.cs:
```diff
@@ -0,0 +1,488 @@
+            }
+        }
+
+        Task IPartitionReceiveHandler.ProcessErrorAsync(Exception error)
```
>ProcessErrorAsync [](start = 38, length = 17)

Do we need to bail out on non-transient errors? This is different than EPH's logic.

JamesBirdsall replied:
EPH must stay running, so it never bails out. With SFP, Service Fabric provides that guarantee for us. On a fatal error, just let it bubble up. User can catch it and apply their own logic in that case, or pass it through to SF and let SF do the restart, possibly on another node.

serkantkaraca commented on src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/EventProcessorService.cs:
```diff
@@ -0,0 +1,488 @@
+                //
+                Exception lastException = null;
+                EventProcessorEventSource.Current.Message("Creating event hub client");
+                for (int i = 0; i < Constants.RetryCount; i++)
```
>for (int i = 0; i < Constants.RetryCount; i++) [](start = 16, length = 46)

Seems we have couple similar retry blocks. Better to move them to a try-catch delegate.

JamesBirdsall replied:
Unified the pattern in a method that takes a delegate.

serkantkaraca committed:
Merge branch 'dev' into SFprocessor

WhitWaldo committed:
GetServiceParitionId operated under the assumption that the full ranged list could be iterated through to count partitions. Replaced with a lookup that doesn't throw when this isn't the case  (e.g. low and high key aren't long.MinValue and long.MaxValue, respectively). (#306)

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Remove IEventProcessorFactory mechanism

JamesBirdsall committed:
Move usings into namespace

JamesBirdsall committed:
Receive timeout option handled by client pump
Do not need to recheck in our event handler

JamesBirdsall committed:
Unify retry pattern

JamesBirdsall committed:
Hooked up receiver options for Event Hubs receiver

JamesBirdsall committed:
Change how settings are done
Make RunAsync throw if called while already running

JamesBirdsall committed:
Provide more ways to set conn str and consumer grp
Handle encrypted config for conn str

JamesBirdsall committed:
Make sure RunAsync can be called again after returning

JamesBirdsall committed:
Rename main class to ServiceFabricProcessor

JamesBirdsall committed:
Move conn str/consumer grp to constructor
Do not retrieve from config

JamesBirdsall committed:
Remove unused constants

JamesBirdsall committed:
Make ServiceFabricProcessor object run-once

JamesBirdsall committed:
Merge branch 'dev' of https://github.com/Azure/azure-event-hubs-dotnet into SFprocessor

JamesBirdsall committed:
Comments improve.
Change to $"" string substitution style.
EventData change to allow setting system props for mocking.

JamesBirdsall committed:
Fix race caused by async lambda with RetryWrapper

JamesBirdsall committed:
Revert EventData to dev

JamesBirdsall committed:
Disable mocking until EventData supports

JamesBirdsall committed:
Update package settings and dependencies

JamesBirdsall committed:
Avoid dependency on StatefulServiceContext class

JamesBirdsall committed:
Fix capitalization of some members

JamesBirdsall committed:
Make consumer group argument required

JamesBirdsall committed:
More capitalization fixes

JamesBirdsall committed:
Sanitize constructor arguments

JamesBirdsall committed:
Changes to work with client 2.2.1

JamesBirdsall committed:
Version 0.5.1 to defeat VS caching

JamesBirdsall committed:
First draft of Programmer's Guide

JamesBirdsall committed:
Tweaks to Programmer's Guide

JamesBirdsall committed:
Fix copyright message, update desc and version

JamesBirdsall committed:
Back out changes to other modules

JamesBirdsall committed:
Back out one more change to another module

JamesBirdsall committed:
Unrelated test oops is blocking merge

sjkwak committed:
Merge branch 'dev' into SFprocessor

serkantkaraca committed:
Merge branch 'dev' into SFprocessor
